### DISCUSSION TITLE
(NEW SCRIPT) Sketcher: Quickly Sketch Drawings on a Node

### DISCUSSION THREAD

**Discussion by i-plasm** [freeplane/freeplane#1496] [Show and tell]
# Sketcher: quickly sketch drawings on a node

**NOTE: Sketcher as a standalone script is now superseded by the Sketcher service bundle within IntelliFlow - an intelligent service-provider menu for Freeplane -. Any updates and maintainance will go toward the [Sketcher IntelliFlow service bundle:](https://github.com/freeplane/freeplane/discussions/1534) https://github.com/freeplane/freeplane/discussions/1534**

Hello everybody! 

I'd like to introduce the ['Sketcher'](https://github.com/i-plasm/freeplane-scripts/wiki/IntelliFlow-Updates) utility. It is a Freeplane IntelliFlow service for quickly sketching drawings on a node. 

Select a node, call the script and a new blank image or an existing one will be opened in a external image editor. Changes to the image will be refreshed in your mindmap. 

![sketcher_23-10-26](https://github.com/freeplane/freeplane/assets/144830522/572ef726-e782-4dc1-be5f-8abf4cf47c47)

## The external viewer

You may or may not specify a viewer. If the viewer is not specified or can not be successfully used, then the script will open the image with the default system image viewer. 

Any viewer should work as long as it understands the syntax **[my_viewer_binary] [image_filepath]** for opening files. Examples of viewers are Gimp, MS Paint, or Drawing.

Instructions on specifying a viewer: https://github.com/i-plasm/freeplane-scripts/wiki/Configuring-Sketcher:-The-external-viewer

## User Templates

Sketcher supports user-defined templates that help customize how sketcher creates images from scratch.

![demo_templates_sketcher](https://github.com/freeplane/freeplane/assets/144830522/3e987071-60f0-4de2-b170-bb7c6ff7cf70)

Users can add as many templates as desired, for instance:

for white background
for grey background
for grid paper
for a blank draw.io-editable image

Instructions for configuring templates: https://github.com/i-plasm/freeplane-scripts/wiki/Configuring-Sketcher:-Templates

## Permissions

The following script permissions are required (set them in Preferences -> Plug-ins): __1- Permit File read; 2- Permit File write; 3- Permit to execute other applications__

## More specs

If a new image was created, it will be stored on the folder 'MyMapName_files', and this folder will be located on the folder containing the mindmap.

 [Get Sketcher as IntelliFlow service bundle:](https://github.com/freeplane/freeplane/discussions/1534) https://github.com/freeplane/freeplane/discussions/1534

Feel free to check for updates and leave feedback! Thanks for the ideas given in this discussion: https://github.com/freeplane/freeplane/discussions/593

**Comment by euu2021** [comment]
It will be very useful! Thanks!

**Comment by i-plasm** [comment]
__Sketcher update 2023-10-28__: 

- It now supports handling images with relative paths, and respects the Freeplane preference for relative vs. absolute path. (Please update the script if your maps are set to use relative paths.)

**Comment by Illegit** [comment]
Nice addon. Useful! Thank you.

**Comment by i-plasm** [comment]
**News 2023-11-16**:
- Sketcher as a standalone script is now superseded by the Sketcher service bundle within IntelliFlow - an intelligent service-provider menu for Freeplane -. Any updates and maintainance will go toward the [Sketcher IntelliFlow service bundle:](https://github.com/freeplane/freeplane/discussions/1534) https://github.com/freeplane/freeplane/discussions/1534

**Update 2023-11-16**:
- Bug fix: if node core text had certain special characters and/or line breaks, Sketcher would fail to create a new blank image

**Comment by Illegit** [comment]
Hi I-plasm I discovered that using Paint 3D it is working only with 2D images but not with 3D images, is it possible to extend it to 3D with Paint 3D?

**Reply by i-plasm** [reply]
Hi, are you editing your PNG in Paint 3D and saving it as a so-called "Paint 3D Project"? If so, that project is saved in some sort of obscure database in a obscure directory, assigning by default the <My_Filename>.PNG title to it, while the PNG itself goes untouched - thus Freeplane won't show any change. If you'd like to launch the 3D project from Freeplane, my suggestion is:

1. When saving ***for the first time*** your Paint 3D project, do it via " Menu -> Save As..." -> "3D Model". This will save it as a 3D project of either `.glb` or `.3mf` extension.
1. Create a node hyperlink that points to the `.glb/.3mf` file. If clicking the hyperlink opens "3D Viewer", it has the option `File -> Open with Paint 3D` 
1. (Optional) it could be helpful to use Sketcher for assigning a screenshot to the node.

**Comment by Illegit** [comment]
Hi I-plasm. I didn't see your answer. Thank you for your explanations, I'll do it.

**Comment by projectresonance** [comment]
HI i-Plasm,

I was trying to get your code to work for draw.io. Specifically, this one is the desktop version:

https://github.com/jgraph/drawio-desktop

Right now it only partially works. If I already make a diagram and save it as a .png it will work as long as I include in a copy of the diagram with it. I don't know what it's doing to make that happen but that is what I notice is the key.

Images to show it working fine for diagram I save with copy:
![image](https://github.com/freeplane/freeplane/assets/51094831/c31d210b-e5ed-49ae-a293-ef335fe2bfa0)

![image](https://github.com/freeplane/freeplane/assets/51094831/2eac2bf5-9ff7-470e-a14e-1dc397a8f0cb)

Images to show if I take a normal .png file with no extra attributes that I'm guessing draw.io adds:

![image](https://github.com/freeplane/freeplane/assets/51094831/3872b65c-a9d9-4d1d-83e9-449b9b24120f)
![image](https://github.com/freeplane/freeplane/assets/51094831/922ffbcc-a41c-49e3-9033-8ba158a5aeab)

Fails and shows empty space which is the same thing that happens when I try to create a new drawing from sketcher on a node. I'm guessing I need to find something to attach to the .png file that your making inside the script so that draw.io can recognize it as editable. Unsure if it could cause errors/be risky but if it's a blank .png file I should hope it wouldn't be bad. Either way, just wondering if you have a suggestion before I dig deep into this.

**Reply by i-plasm** [reply]
Hi @projectresonance ,

I'm not a user of draw.io, but I have a few suggestions.

In [this link](https://www.drawio.com/doc/faq/save-file-formats) it says that if you go `File > Save As... > PNG` then it *"saves your diagram as a PNG image with embedded diagram data"*. So indeed, the diagram is included as meta-info to the PNG. Therefore using PNGs generated by other apps won't work. (Note: the diagram is also embedded when you produce it via [exporting to PNG](https://www.drawio.com/doc/faq/export-to-png) ***as long as***  the option "Include a copy of my diagram" is checked)

> I try to create a new drawing from sketcher on a node... I'm guessing I need to find something to attach to the .png file that your making inside the script so that draw.io can recognize it as editable.

If you'd like to use Sketcher to create such draw.io-editable PNGs from scratch, a solution is making a new option in Sketcher that leads it to create a copy of a "blank" draw.io-editable PNG when you want a new drawing on a node. In other words, modifying Sketcher to support templates. 

> Unsure if it could cause errors/be risky but if it's a blank .png file  I should hope it wouldn't be bad.

The solution I mentioned should not pose any type of risk. However the risk could come from the human factor **^^IF^^** you keep both .PNG files and other extension(s) for the same diagram: in such case every time you decide to use Sketcher to edit an existing PNG, you would have to make sure both the PNG and the.XML/.DRAWIO/.SVG files indeed represent the same diagram, otherwise your PNG might represent an outdated version. My suggestion to totally avoid this risk would be to **exclusively** rely on PNGs, and to not maintain .DRAWIO/.XMLs/etc (so that you never have to worry about the issue of syncing). Of course whether all that's possible would depend on your workflow. 

Also, **it is really important to avoid editing the PNG with any software other than draw.io <s>unless you are sure it will not strip the PNG from its meta-info, otherwise thus you'd lose the embedded diagram</s> because draw.io won't take those changes into account and will overwrite them. In other words, during the entire lifetime of the PNG, the only software that can be used to edit it should be draw.io.** Finally, if the PNG is uploaded to certain webservices such as Whatsapp, Google Photos, etc, the embedded diagram may be lost since those services usually process and diminish the quality of images. So they should not be used if the embedded diagram is to be preserved.

Let me know if the proposed solution is of your interest.

**Reply by projectresonance** [reply]
HI i-plasm,

Yes that would be very nice! Figuring out a way to bring in a .png already on a node that doesn't have the data into a new diagram would be nice if that doesn't take too much time.(i.e.  create a copy of a blank draw.io editable PNG, take  .png in the node and place it in the diagram, and once done editing it just replaces that .png saved on the node with the new one created from draw.io). In some cases, I find I might need to highlight some features on an image and I have a map with some .pngs already on there that I would like to highlight some points on. However,  copying images into a new blank diagram myself and clearing the nodes isn't that bad so it's not really that important.

So that solution you proposed is of interest.

**Reply by i-plasm** [reply]
Hi @projectresonance ,

> Figuring out a way to bring in a .png already on a node that doesn't have the data into a new diagram would be nice

This is basically a direct converter of PNG to draw.io-editable PNG. That should possible with enough understanding of the draw.io format and how it embeds images. It seems approachable.

> that solution you proposed is of interes

I've implemented the Templates feature in Sketcher. It's available in the latest IntelliFlow version. I added a wiki on Templates with explanations: https://github.com/i-plasm/freeplane-scripts/wiki/Configuring-Sketcher:-Templates

I tested it in a Windows machine and it works fine, but please test it with dummy maps and images first to make sure it works well for you. Let me know if you try it out. (P.S I installed draw.io and noticed that the PNGs should never be edited with any other software because draw.io won't take those changes into account and will overwrite them. In other words, during the entire lifetime of the draw.io-editable PNG, the only software that can be used to edit it must be draw.io. Also, avoid image-sharing web services such as whatsapp, google photos, etc, since those usually compress/lower the size of PNGs and more than likely will delete the meta-info.)

**Reply by projectresonance** [reply]
Hi @i-plasm, 

I've been testing the program between last week and this week. So far, it's been holding up well and has opened up quite some more flexibility in diagramming so I must thank you for that. I'll see if I run into any issues or think of another idea. I almost think the Sketcher should come with Freeplane, especially since one of the main mind mapping software(I believe Mindomo) has a built in drawing tool and I had found some use out of it. 

But rambling aside, thank you. This has been helpful.
